src/Database/LSMTree/Internal.hs

```diff
@@ -1278,3 +1306,140 @@ duplicate t@Table{..} = do
             tableConfig
             tableArenaManager
             content
+
+
+{-------------------------------------------------------------------------------
+   Table union
+-------------------------------------------------------------------------------}
+
+{-# SPECIALISE union :: Table IO h -> Table IO h -> IO (Table IO h) #-}
+-- | See 'Database.LSMTree.Normal.union'.
+union ::
+     (MonadMask m, MonadMVar m, MonadST m, MonadSTM m)
+  => Table m h
+  -> Table m h
+  -> m (Table m h)
+union t1 t2 = unions $ V.fromList [t1, t2]
+
+{-# SPECIALISE unions :: V.Vector (Table IO h) -> IO (Table IO h) #-}
+-- | See 'Database.LSMTree.Normal.unions'.
+unions ::
+     (MonadMask m, MonadMVar m, MonadST m, MonadSTM m)
+  => V.Vector (Table m h)
+  -> m (Table m h)
+unions ts
+  | n == 0 = throwIO ErrUnionsZeroTables
+  | otherwise = do
+      conf <-
+        case vmatch (V.map tableConfig ts) of
+          Left (i, j) -> throwIO $ ErrUnionsTableConfigMismatch i j
+          Right conf  -> pure conf
+
+      sesh <-
+        vmatchSessions ts >>= \case
+          Left (i, j) -> throwIO $ ErrUnionsSessionMismatch i j
+          Right sesh -> pure sesh
+
+      traceWith (sessionTracer sesh) $ TraceUnions n
+
+      -- We acquire a read-lock on the session open-state to prevent races, see
+      -- 'sessionOpenTables'.
+      modifyWithTempRegistry
+        (atomically $ RW.unsafeAcquireReadAccess (sessionState sesh))
+        (\_ -> atomically $ RW.unsafeReleaseReadAccess (sessionState sesh)) $ \reg -> \case
+          SessionClosed -> throwIO ErrSessionClosed
+          seshState@(SessionOpen seshEnv) -> do
+            contents <-
+              V.forM ts $ \t -> do
+                withOpenTable t $ \tEnv ->
+                  -- The table contents escape the read access, but we just added references
+                  -- to each run so it is safe.
+                  RW.withReadAccess (tableContent tEnv) (duplicateTableContent reg)
+
+            content <-
+              error "unions: combine contents into merging tree" $ -- TODO
+                contents
+
+            t <-
+              newWith
+                reg
+                sesh
+                seshEnv
+                conf
+                (error "unions: ArenaManager") -- TODO
+                content
+
+            pure (seshState, t)
+  where
+    n = V.length ts
+
+-- | Like 'vmatchBy', but the match function is @(==)@.
+vmatch :: Eq a => V.Vector a -> Either (Int, Int) a
+vmatch = vmatchBy (==)
+
+-- | Check that all values in the vector match. If so, return the matched value.
+-- If there is a mismatch, return the vector indices of the mismatching values.
+--
+-- Assumes the input vector is non-empty.
+vmatchBy :: (a -> a -> Bool) -> V.Vector a -> Either (Int, Int) a
+vmatchBy cmp xs0 =
```

Comment:
Just a nit, but `cmp` reminds me of `compare :: a -> a -> Ordering`. Maybe something like `eq`?